<a href="https://colab.research.google.com/github/Ariely13/Tugas/blob/main/ArilMuhammadNurham_241061012_Tugas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Aril_Muhammad_Nurham_241061012
## Tugas Ke 2 - Teknik Pengambilan Sampel dan Data Wrangling

In [9]:
# mengimpor library yang diperlukan
import pandas as pd
import numpy as np

In [10]:
# baca dataset
df = pd.read_csv('/content/data_tugas 2.csv')

# ganti tanda tanya dengan NaN karena merupakan missing value
df.replace('?', np.nan, inplace=True)

print(df.head())
print(df.shape)

  normalized-losses         make bahan-bakar jumlah-pintu   body-style  \
0               NaN  alfa-romero         gas          two  convertible   
1               NaN  alfa-romero         gas          two  convertible   
2               NaN  alfa-romero         gas          two    hatchback   
3               164         audi         gas         four        sedan   
4               164         audi         gas         four        sedan   

   wheel-bases  length  width  height  
0         88.6   168.8   64.1    48.8  
1         88.6   168.8   64.1    48.8  
2         94.5   171.2   65.5    52.4  
3         99.8   176.6   66.2    54.3  
4         99.4   176.6   66.4    54.3  
(205, 9)


Berikut versi yang lebih rapi dan enak dibaca:

Kode tersebut digunakan untuk menginisialisasi dataset dengan membaca file CSV serta melakukan pembersihan data awal. Nilai berupa tanda tanya (`?`) diubah menjadi `NaN` (*Not a Number*) agar dikenali sebagai *missing value* oleh library Pandas dan dapat diproses lebih lanjut. Berdasarkan output `df.shape`, dataset memiliki 205 baris dan 9 kolom.


#1. Menampilkan jumlah missing value

In [11]:
missing = df.isnull().sum()
print("\nJumlah missing value tiap kolom:") # menghitung jumlah nilai kosong (NaN) di setiap kolom
print(missing)

# cari kolom yang paling banyak missing valuenya
print("\nKolom dengan missing value terbanyak:")
print(missing[missing > 0].sort_values(ascending=False))
# menampilkan hanya kolom yang masih memiliki missing value, lalu diurutkan dari yang terbanyak


Jumlah missing value tiap kolom:
normalized-losses    41
make                  4
bahan-bakar          13
jumlah-pintu         18
body-style           12
wheel-bases           3
length                8
width                 5
height                4
dtype: int64

Kolom dengan missing value terbanyak:
normalized-losses    41
jumlah-pintu         18
bahan-bakar          13
body-style           12
length                8
width                 5
make                  4
height                4
wheel-bases           3
dtype: int64


Kode ini digunakan untuk mengidentifikasi sekaligus mengurutkan sebaran nilai yang hilang (*missing values*) dalam dataset. Fungsi `isnull().sum()` menghitung jumlah data kosong pada setiap kolom, kemudian `sort_values()` mengurutkannya dari yang terbanyak hingga paling sedikit. Hasil yang diperoleh menunjukkan bahwa kolom *normalized-losses* memiliki jumlah missing value paling tinggi, yaitu sebanyak 41 data, sehingga perlu menjadi prioritas utama dalam proses pembersihan data agar tidak memengaruhi hasil analisis.


# 2. Ganti missing value pada variabel kategori

In [12]:
# kolom kategori yang ada missing valuenya
kolom_kategori = ['make', 'bahan-bakar', 'jumlah-pintu', 'body-style']

# isi missing value dengan 'unknown'
for col in kolom_kategori: # looping untuk setiap kolom kategori
    df[col].fillna('unknown', inplace=True) # mengisi nilai kosong dengan 'unknown'

# cek hasilnya
print("\nRingkasan kategori setelah diisi unknown:")  # menampilkan judul output
for col in kolom_kategori:                            # looping lagi untuk melihat hasil
    print(f"\n{col}:")                                # menampilkan nama kolom
    print(df[col].value_counts())                     # menampilkan jumlah tiap kategori


Ringkasan kategori setelah diisi unknown:

make:
make
toyota           32
mazda            17
nissan           17
honda            13
mitsubishi       12
volkswagen       12
volvo            11
subaru           11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           4
unknown           4
isuzu             4
alfa-romero       3
chevrolet         3
jaguar            3
renault           2
mercury           1
Name: count, dtype: int64

bahan-bakar:
bahan-bakar
gas        173
diesel      19
unknown     13
Name: count, dtype: int64

jumlah-pintu:
jumlah-pintu
four       101
two         86
unknown     18
Name: count, dtype: int64

body-style:
body-style
sedan          88
hatchback      67
wagon          24
unknown        12
hardtop         8
convertible     6
Name: count, dtype: int64


/tmp/ipykernel_1315/2326405222.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna('unknown', inplace=True) # mengisi nilai kosong dengan 'unknown'


Kode tersebut digunakan untuk menangani missing value pada kolom kategori dalam dataset. Sebuah list dibuat untuk menentukan kolom yang akan diproses, kemudian dilakukan perulangan untuk mengisi nilai kosong (NaN) dengan `'unknown'` menggunakan fungsi `fillna()` agar data lebih konsisten. Setelah itu, ditampilkan ringkasan setiap kolom menggunakan `value_counts()` untuk melihat distribusi data setelah proses pembersihan.


# 3. Ganti missing value pada variabel numerik pakai mean

In [13]:
# kolom numerik yang perlu diimputasi
kolom_numerik = ['length', 'width', 'height']

# pastikan tipenya numerik dulu
for col in kolom_numerik: # looping untuk setiap kolom numerik
    df[col] = pd.to_numeric(df[col], errors='coerce') # ubah ke tipe numerik, nilai tidak valid jadi NaN

# hitung mean tiap kolom terus isi yang kosong
for col in kolom_numerik: # looping lagi untuk proses imputasi
    rata2 = df[col].mean() # menghitung rata-rata tiap kolom
    print(f"\nMean {col}: {rata2:.2f}") # menampilkan nilai rata-rata
    df[col].fillna(rata2, inplace=True) # mengisi nilai kosong dengan rata-rata

print("\ncek missing value setelah diisi:")
print(df[kolom_numerik].isnull().sum())


Mean length: 173.90

Mean width: 65.93

Mean height: 53.74

cek missing value setelah diisi:
length    0
width     0
height    0
dtype: int64


/tmp/ipykernel_1315/2740012568.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(rata2, inplace=True) # mengisi nilai kosong dengan rata-rata


Bagian ini digunakan untuk mengisi nilai kosong pada kolom numerik dalam dataset. Pertama, ditentukan kolom yang akan diproses yaitu `length`, `width`, dan `height`. Kemudian, setiap kolom diubah ke tipe numerik agar bisa dilakukan perhitungan, dan jika ada data yang tidak valid akan otomatis dianggap sebagai nilai kosong (NaN). Setelah itu, dihitung nilai rata-rata (mean) dari masing-masing kolom, lalu nilai yang kosong diisi menggunakan rata-rata tersebut. Terakhir, dilakukan pengecekan untuk memastikan bahwa tidak ada lagi nilai kosong yang tersisa pada kolom-kolom tersebut.


# 4. Min-Max Normalization pada wheel-bases

In [14]:
# pastikan numerik
df['wheel-bases'] = pd.to_numeric(df['wheel-bases'], errors='coerce')

# rumus min max = (x - min) / (max - min)
min_wb = df['wheel-bases'].min() # mengambil nilai minimum dari kolom wheel-bases
max_wb = df['wheel-bases'].max() # mengambil nilai maksimum dari kolom wheel-bases

df['wheel-bases-normalized'] = (df['wheel-bases'] - min_wb) / (max_wb - min_wb)
# melakukan normalisasi min-max agar nilai berada di rentang 0–1 dan menyimpannya ke kolom baru

print("\nHasil min-max normalization wheel-bases:")
print(df[['wheel-bases', 'wheel-bases-normalized']].head(10))


Hasil min-max normalization wheel-bases:
   wheel-bases  wheel-bases-normalized
0         88.6                0.058309
1         88.6                0.058309
2         94.5                0.230321
3         99.8                0.384840
4         99.4                0.373178
5         99.8                0.384840
6        105.8                0.559767
7        105.8                0.559767
8        105.8                0.559767
9         99.5                0.376093


Setelah itu untuk melakukan normalisasi data pada kolom wheel-bases menggunakan metode min-max normalization. Pertama, data pada kolom tersebut diubah ke tipe numerik agar bisa dihitung, dan jika ada nilai yang tidak valid akan dianggap sebagai NaN. Kemudian, diambil nilai minimum dan maksimum dari kolom tersebut. Setelah itu, setiap nilai diubah menggunakan rumus (x - min) / (max - min) sehingga hasilnya berada pada rentang 0 sampai 1. Nilai yang sudah dinormalisasi disimpan pada kolom baru bernama wheel-bases-normalized. Terakhir, ditampilkan beberapa baris data untuk melihat hasil normalisasi yang telah dilakukan.

# 5. Z-Score Normalization pada length, width, height

In [15]:
# rumus z-score = (x - mean) / std
kolom_zscore = ['length', 'width', 'height']

for col in kolom_zscore:  # looping untuk setiap kolom dalam daftar kolom_zscore
    mean = df[col].mean() # menghitung nilai rata-rata (mean) dari kolom
    std = df[col].std()  # menghitung standar deviasi (std) dari kolom
    df[col + '-zscore'] = (df[col] - mean) / std  # menghitung Z-score dan menyimpannya ke kolom baru
    print(f"\nZ-score {col} selesai, mean={mean:.2f}, std={std:.2f}") # menampilkan hasil proses tiap kolom

print("\nHasil z-score normalization:")
print(df[['length', 'length-zscore', 'width', 'width-zscore', 'height', 'height-zscore']].head())


# simpan hasil
df.to_csv('data_tugas_2_cleaned.csv', index=False)
print("\nselesai, file disimpan")




Z-score length selesai, mean=173.90, std=11.91

Z-score width selesai, mean=65.93, std=2.11

Z-score height selesai, mean=53.74, std=2.42

Hasil z-score normalization:
   length  length-zscore  width  width-zscore  height  height-zscore
0   168.8      -0.428526   64.1     -0.867630    48.8      -2.038094
1   168.8      -0.428526   64.1     -0.867630    48.8      -2.038094
2   171.2      -0.226967   65.5     -0.204232    52.4      -0.552035
3   176.6       0.226541   66.2      0.127467    54.3       0.232274
4   176.6       0.226541   66.4      0.222238    54.3       0.232274

selesai, file disimpan


Bagian terakhir untuk melakukan **normalisasi Z-score** pada kolom `length`, `width`, dan `height`. Untuk setiap kolom, dihitung nilai rata-rata (mean) dan standar deviasi (std), kemudian setiap data diubah dengan rumus *(x - mean) / std* agar memiliki skala yang lebih seimbang. Hasilnya disimpan pada kolom baru dengan tambahan `-zscore`. Setelah itu, ditampilkan beberapa data untuk melihat hasil normalisasi. Terakhir, dataset yang sudah dibersihkan dan diproses disimpan ke file baru bernama `data_tugas_2_cleaned.csv`.
